std-bits/table/src/main/java/org/enso/table/data/table/Table.java

```diff
@@ -219,64 +216,55 @@ public Table selectColumns(List<String> colNames) {
    */
   public Table join(Table right, List<JoinCondition> conditions, boolean keepLeftUnmatched, boolean keepMatched, boolean keepRightUnmatched, boolean includeLeftColumns, boolean includeRightColumns, List<String> rightColumnsToDrop, String right_prefix, Comparator<Object> objectComparator, BiFunction<Object, Object, Boolean> equalityFallback) {
     NameDeduplicator nameDeduplicator = new NameDeduplicator();
-    JoinResult joinResult = null;
-    // Only compute the join if there are any results to be returned.
-    if (keepLeftUnmatched || keepMatched || keepRightUnmatched) {
-      JoinStrategy strategy = new IndexJoin(objectComparator, equalityFallback);
-      joinResult = strategy.join(this, right, conditions);
+    if (!keepLeftUnmatched && !keepMatched && !keepRightUnmatched) {
+      throw new IllegalArgumentException("At least one of keepLeftUnmatched, keepMatched or keepRightUnmatched must be true.");
     }
 
-    List<Integer> leftRows = new ArrayList<>();
-    List<Integer> rightRows = new ArrayList<>();
+    JoinStrategy strategy = new IndexJoin(objectComparator, equalityFallback);
+    JoinResult joinResult = strategy.join(this, right, conditions);
+
+    List<JoinResult> resultsToKeep = new ArrayList<>();
 
     if (keepMatched) {
-      for (var match : joinResult.matchedRows()) {
-        leftRows.add(match.getLeft());
-        rightRows.add(match.getRight());
-      }
+      resultsToKeep.add(joinResult);
     }
 
     if (keepLeftUnmatched) {
-      HashSet<Integer> matchedLeftRows = new HashSet<>();
-      for (var match : joinResult.matchedRows()) {
-        matchedLeftRows.add(match.getLeft());
-      }
-
+      Set<Integer> matchedLeftRows = joinResult.leftMatchedRows();
+      JoinResult.Builder leftUnmatchedBuilder = new JoinResult.Builder();
       for (int i = 0; i < this.rowCount(); i++) {
         if (!matchedLeftRows.contains(i)) {
-          leftRows.add(i);
-          rightRows.add(Index.NOT_FOUND);
+          leftUnmatchedBuilder.addRow(i, Index.NOT_FOUND);
```

Comment:
`resultsToKeep` is an aggregator for `JoinResults` that later will be merged.

The idea is that the `JoinResult.Builder` is building an array of primitives (to avoid unnecessary boxing of integers).

I guess I could rewrite this to use a single `JoinResult.Builder` for the three sub-operations, if you think it is worth it.